<a href="https://colab.research.google.com/github/IGiovanniRamirez/Giovanni-Ramirez-Data-Portfolio/blob/main/An%C3%A1lisis%20e-commerce%20Dropshipping/An%C3%A1lisis_de_comercio_electr%C3%B3nico_y_dropshipping.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🧠 Dropshipping Profitability Analytics  
### End-to-End Profit Optimization (Operación + Marketing) en modelo Cash-on-Delivery (COD)

---

## 📌 1. Contexto de negocio

Este proyecto se basa en datos reales de una operación de e-commerce bajo el modelo de **dropshipping con pago contra entrega (COD)** en Colombia.

El negocio operaba a través de:

- **Shopify** → canal comercial (tienda online)  
- **Dropi** → operación logística y financiera  
- **Meta Ads (Facebook / Instagram) y TikTok Ads** → adquisición de clientes  

A diferencia de análisis tradicionales, este proyecto incorpora una visión **end-to-end**, integrando:

- Ingresos  
- Costos de producto  
- Costos logísticos  
- Costos de marketing  

Esto permite analizar la **rentabilidad real del negocio**, más allá de métricas superficiales.

---

### ⚠️ Desafíos del modelo COD

- Alto riesgo de pedidos no efectivos (cancelaciones y devoluciones)  
- Costos logísticos elevados (envíos fallidos y devoluciones)  
- Dependencia de validación manual de pedidos  
- El marketing se paga incluso si el pedido no se entrega  

---

## 🔁 2. Proceso operativo del negocio

Flujo real de la operación:

1. Cliente ve anuncio (Meta Ads / TikTok Ads)  
2. Cliente realiza pedido en Shopify  
3. Confirmación automática vía chatbot (Dropi)  
4. Validación manual si no hay respuesta, cancelación de pedido en caso de tampoco encontrar respuesta en la validación manual.  
5. Procesamiento logístico  
6. Resultado del pedido:

- ✅ **ENTREGADO** → venta exitosa  
- ⚠️ **DEVOLUCIÓN / DEVUELTA** → pérdida logística  
- ❌ **CANCELADO / RECHAZADO** → pedido no ejecutado  
- ⚫ **RECOGIDA FALLIDA** → pedido no procesado  
- 🔴 **INDEMNIZADA POR DROPI** → pérdida total (producto + envío + reembolso)  

---

## 🎯 3. Objetivo del proyecto

Analizar la rentabilidad del negocio de forma integral, identificando:

- Pérdidas operativas  
- Ineficiencias logísticas  
- Impacto del marketing en los resultados  
- Relación entre adquisición de clientes y rentabilidad  

---

## 📂 4. Fuentes de datos

Se utilizan tres datasets principales:

### 📦 Operación (Dropi)
- `ordenes_productos_raw` → pedidos y productos (nivel línea)  
- `clientes_raw` → información de clientes  

### 📣 Marketing
- `marketing_raw` → inversión en campañas publicitarias  
  - Meta Ads  
  - TikTok Ads  
  - Shopify (tracking de gasto asociado)

---

### ⚠️ Decisión sobre datos de Shopify

Los datasets de Shopify fueron descartados porque:

- No incluían costos de producto  
- No incluían costos logísticos  
- Información incompleta de clientes  
- No reflejaban el estado real de la operación  

Se prioriza **Dropi** por permitir un análisis financiero completo.

---

## 💰 5. Alcance del análisis

### ✔ Incluye:
- Revenue  
- Costos de producto (COGS)  
- Costos logísticos  
- Costos de marketing  
- Profit operativo  
- Profit neto  

---

### 📊 Niveles de análisis

- **Profit Operativo** → sin marketing  
- **Profit Neto** → incluyendo marketing  

---

### ⚠️ Implicación clave

El análisis permite responder:

> ¿El negocio realmente gana dinero después de adquirir clientes?

---

## 🏗️ 6. Arquitectura del proyecto


Raw Data (Dropi + Marketing Ads)  
↓  
Data Cleaning (Python - Pandas)  
↓  
Data Modeling (SQL - enfoque práctico)  
↓  
Data Integration (tabla final unificada)  
↓  
Exploratory Data Analysis (EDA)  
↓  
KPI Calculation (Operativos + Marketing)  
↓  
Dashboard (Tableau)  
↓  
Insights & Business Recommendations


---

## 🧩 7. Definición del proyecto

El proyecto analiza el negocio desde una perspectiva integral:

- Operación  
- Logística  
- Marketing  
- Rentabilidad  

---

## ❗ 8. Problema de negocio

En el modelo COD:

- No todos los pedidos generan ingresos  
- Algunos pedidos generan pérdidas directas  
- El marketing es un costo incluso si el pedido falla  

---

### 🎯 Pregunta clave

> ¿El negocio es rentable después de considerar TODOS los costos?

---

## ❓ 9. Preguntas de negocio

### 💰 Rentabilidad
- ¿Cuál es la ganancia neta del negocio?  
- ¿Cuál es el margen real?  
- ¿Qué proporción de pedidos genera pérdidas?  

---

### 📣 Marketing
- ¿Cuál es el costo por pedido?  
- ¿Cuál es el costo por pedido entregado?  
- ¿El marketing está afectando la rentabilidad?  

---

### 🚚 Logística
- ¿Qué tipo de pedidos generan mayor pérdida?  
- ¿Cuál es el impacto de las devoluciones?  

---

### ⚠️ Riesgo operativo
- ¿Qué estados destruyen valor?  
- ¿Qué porcentaje de pedidos no es rentable?  

---

### 📦 Productos
- ¿Qué productos son rentables?  
- ¿Cuáles generan pérdidas recurrentes?  

---

## 📊 10. KPIs del proyecto

### 💰 Financieros
- Revenue  
- Cost Product  
- Cost Shipping  
- Cost Return  
- Marketing Cost  
- Profit Operativo  
- Profit Neto  
- Profit Margin  

---

### 📣 Marketing
- Marketing Cost per Order  
- Marketing Cost per Delivered Order  

---

### 🚚 Operativos
- Fulfillment Rate  
- Return Rate  
- Cancellation Rate  

---

### ⚠️ Riesgo
- Loss Rate  

---

## 🧠 11. Enfoque analítico

### 1. Análisis end-to-end
Se evalúa todo el flujo:
> adquisición → operación → resultado financiero  

---

### 2. Doble capa de rentabilidad
- Operativa  
- Neta (con marketing)  

---

### 3. Análisis por estado del pedido
- Delivered  
- Returned  
- Cancelled  
- Critical Loss  

---

### 4. Evaluación del impacto del marketing
Se analiza:

- Costo por pedido  
- Costo por pedido entregado  
- Impacto en profit  

---

## 🧱 12. Modelo de datos (implementado)

El modelo final es una **tabla denormalizada optimizada para BI**:

### ⭐ Tabla final: `fact_final`

Granularidad:
> Una fila por producto dentro de cada pedido  

---

### 📌 Columnas clave

- order_id  
- fecha  
- product  
- city  
- department  
- status_group  
- revenue  
- cost_product  
- cost_shipping  
- cost_return  
- marketing_cost_allocated  
- profit_operativo  
- profit_neto  

---

### 🚩 Flags
- is_delivered  
- is_returned  
- is_cancelled  
- is_loss  

---

## 📊 13. Estructura del dashboard

### 1. Executive Overview
- Profit neto  
- Margen  
- Distribución de pedidos  

---

### 2. Unit Economics
- Revenue vs Costos  
- Profit por pedido  

---

### 3. Operación
- Fulfillment Rate  
- Return Rate  
- Costos logísticos  

---

### 4. Productos
- Productos rentables  
- Productos con pérdidas  

---

### 5. Marketing
- Costo total  
- Costo por pedido  
- Impacto en profit  

---

## 🚀 14. Insight esperado

Ejemplo:

> “Un porcentaje significativo de pedidos no solo no genera ingresos, sino que produce pérdidas debido a costos logísticos. Al incorporar los costos de marketing, se evidencia que parte del negocio no es rentable, especialmente en pedidos no entregados y productos con alta tasa de devolución.”

# Raw Data (Dropi + Marketing Ads) 👇

In [1]:
# Carga inicial de datasets (Python)
import pandas as pd

orders_products_raw = pd.read_excel("https://drive.google.com/uc?export=download&id=1te_K3eOOdxevAIVxSZQhneO3bbOHd7PK")
customers_raw = pd.read_excel("https://drive.google.com/uc?export=download&id=18IL0kkuGoUsreuRUpVoQjI2rKtc73Uxk")
marketing_raw = pd.read_excel("https://drive.google.com/uc?export=download&id=1Lc_wB0N4ad3dtfJSCd7sSWEYUYC_Uvdz")


In [2]:
# Explorar datasets (Python)
# Explorar la estructura de orders_products_raw
display("Pedidos y productos crudo: vista general de la tabla", orders_products_raw.head())
print(f"\n")
display("Pedidos y productos crudo: información general del dataset", orders_products_raw.info())
print(f"\n")
# Explorar columnas numéricas de orders_products_raw
display("Pedidos y productos crudo: estadísticas de columnas numéricas", orders_products_raw.describe())
print(f"\n")
# Explorar columnas categóricas de orders_products_raw
categoricas_orders_products_raw = ['ID','NOMBRE CLIENTE','TELÉFONO','EMAIL','TIPO DE IDENTIFICACION','NÚMERO GUIA','ESTATUS','TIPO DE ENVIO','DEPARTAMENTO DESTINO','CIUDAD DESTINO','DIRECCION','NOTAS','TRANSPORTADORA','SKU','VARIACION ID','PRODUCTO','VARIACION','NOVEDAD','FUE SOLUCIONADA LA NOVEDAD','SOLUCIÓN','OBSERVACIÓN','ÚLTIMO MOVIMIENTO','CONCEPTO ÚLTIMO MOVIMIENTO','UBICACIÓN DE ÚLTIMO MOVIMIENTO','VENDEDOR','TIPO DE TIENDA','TIENDA','ID DE ORDEN DE TIENDA','NUMERO DE PEDIDO DE TIENDA','TAGS','CONTADOR DE INDEMNIZACIONES','CONTADOR DE INDEMNIZACIONES','CONCEPTO ÚLTIMA INDENMIZACIÓN']
for col in categoricas_orders_products_raw:
  print(f"\n")
  display("Pedidos y productos crudo: estadísticas de columnas categóricas",orders_products_raw[col].value_counts())
  print(f"\n")
# Explorar la estructura de customers_raw
display("Customers crudo: vista general de la tabla", customers_raw.head())
print(f"\n")
display("Customers crudo: información general del dataset", customers_raw.info())
print(f"\n")
# Explorar columnas numéricas de customers_raw
display("Customers crudo: estadísticas de columnas numéricas", customers_raw.describe())
print(f"\n")
# Explorar columnas categóricas de customers_raw
categoricas_customers_raw = ['ID','NAME','LASTNAME','PHONE','EMAIL','ADDRESS1','ADDRESS2','IDENTIFICATION']
for col in categoricas_customers_raw:
  print(f"\n")
  display("Customers crudo: estadísticas de columnas categóricas",customers_raw[col].value_counts())
  print(f"\n")
  # Explorar la estructura de marketing_raw
display("Marketing crudo: vista general de la tabla", marketing_raw.head())
print(f"\n")
display("Marketing crudo: información general del dataset", marketing_raw.info())
print(f"\n")
# Explorar columnas numéricas de marketing_raw
display("Marketing crudo: estadísticas de columnas numéricas", marketing_raw.describe())
print(f"\n")
# Explorar columnas categóricas de marketing_raw
categoricas_marketing_raw = ['Año','Mes','Canal']
for col in categoricas_marketing_raw:
  print(f"\n")
  display("Marketing crudo: estadísticas de columnas categóricas",marketing_raw[col].value_counts())
  print(f"\n")

'Pedidos y productos crudo: vista general de la tabla'

,ID,HORA,FECHA,NOMBRE CLIENTE,TELÉFONO,EMAIL,TIPO DE IDENTIFICACION,NRO DE IDENTIFICACION,NÚMERO GUIA,ESTATUS,...,UBICACIÓN DE ÚLTIMO MOVIMIENTO,VENDEDOR,TIPO DE TIENDA,TIENDA,ID DE ORDEN DE TIENDA,NUMERO DE PEDIDO DE TIENDA,TAGS,FECHA GUIA GENERADA,CONTADOR DE INDEMNIZACIONES,CONCEPTO ÚLTIMA INDENMIZACIÓN
0,20811983,12:03,27-06-2024,Ana Elvinia Bogotá,3133118203,arcillalauravalentina26@gmail.com,NaN,NaN,014139413102,ENTREGADO,...,ESTADO,NaN,SHOPIFY,Casa Práctica Market,5.291478e+12,1038.0,NaN,28/06/2024,NaN,NaN
1,20768332,14:52,26-06-2024,Yury Carolinas Quimbaya,3133320953,jeidisaavedra74@gmail.com,NaN,NaN,014139380655,ENTREGADO,...,ESTADO,NaN,SHOPIFY,Casa Práctica Market,5.290290e+12,1037.0,NaN,27/06/2024,NaN,NaN
2,20768332,14:52,26-06-2024,Yury Carolinas Quimbaya,3133320953,jeidisaavedra74@gmail.com,NaN,NaN,014139380655,ENTREGADO,...,ESTADO,NaN,SHOPIFY,Casa Práctica Market,5.290290e+12,1037.0,NaN,27/06/2024,NaN,NaN
3,20763766,13:20,26-06-2024,Neiser Murillo,3177081685,neisermurillo771@gmail.com,NaN,NaN,NaN,CANCELADO,...,NaN,NaN,SHOPIFY,Casa Práctica Market,5.290208e+12,1036.0,NaN,NaN,NaN,NaN
4,20763766,13:20,26-06-2024,Neiser Murillo,3177081685,neisermurillo771@gmail.com,NaN,NaN,NaN,CANCELADO,...,NaN,NaN,SHOPIFY,Casa Práctica Market,5.290208e+12,1036.0,NaN,NaN,NaN,NaN




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1022 entries, 0 to 1021
Data columns (total 52 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   ID                              1022 non-null   int64  
 1   HORA                            1022 non-null   object 
 2   FECHA                           1022 non-null   object 
 3   NOMBRE CLIENTE                  1022 non-null   object 
 4   TELÉFONO                        1022 non-null   int64  
 5   EMAIL                           33 non-null     object 
 6   TIPO DE IDENTIFICACION          0 non-null      float64
 7   NRO DE IDENTIFICACION           0 non-null      float64
 8   NÚMERO GUIA                     676 non-null    object 
 9   ESTATUS                         1022 non-null   object 
 10  TIPO DE ENVIO                   1022 non-null   object 
 11  DEPARTAMENTO DESTINO            1022 non-null   object 
 12  CIUDAD DESTINO                  

'Pedidos y productos crudo: información general del dataset'

None

'Pedidos y productos crudo: estadísticas de columnas numéricas'

,ID,TELÉFONO,TIPO DE IDENTIFICACION,NRO DE IDENTIFICACION,TOTAL DE LA ORDEN,GANANCIA,PRECIO FLETE,COSTO DEVOLUCION FLETE,COMISION,% COMISION DE LA PLATAFORMMA,...,PRECIO PROVEEDOR X CANTIDAD,PRODUCTO ID,VARIACION ID,CANTIDAD,OBSERVACIÓN,ID DE ORDEN DE TIENDA,NUMERO DE PEDIDO DE TIENDA,TAGS,CONTADOR DE INDEMNIZACIONES,CONCEPTO ÚLTIMA INDENMIZACIÓN
count,1.022000e+03,1.022000e+03,0.0,0.0,1022.000000,596.000000,1022.000000,1022.000000,1022.0,1022.0,...,1022.000000,1.022000e+03,843.000000,1022.000000,0.0,7.400000e+02,740.000000,0.0,0.0,0.0
mean,2.511478e+07,3.139007e+09,NaN,NaN,79878.897769,32770.838104,10523.049100,934.159423,0.0,0.0,...,36775.332681,5.556817e+05,471717.674970,1.083170,NaN,5.392850e+12,1327.131081,NaN,NaN,NaN
std,1.817818e+06,1.816806e+08,NaN,NaN,25883.989701,13447.548734,4676.626908,4587.585127,0.0,0.0,...,16564.129501,1.387598e+05,206689.164588,0.362182,NaN,5.055472e+10,165.284199,NaN,NaN,NaN
min,1.771991e+07,3.001217e+09,NaN,NaN,16100.000000,0.000000,2151.020000,0.000000,0.0,0.0,...,4000.000000,3.095700e+04,66290.000000,1.000000,NaN,5.202434e+12,1004.000000,NaN,NaN,NaN
25%,2.501435e+07,3.104059e+09,NaN,NaN,66950.130000,23743.812500,6901.200000,0.000000,0.0,0.0,...,33000.000000,5.470000e+05,358166.000000,1.000000,NaN,5.352543e+12,1182.750000,NaN,NaN,NaN
50%,2.560684e+07,3.133220e+09,NaN,NaN,74320.000000,29905.510000,8963.220000,0.000000,0.0,0.0,...,36000.000000,5.474140e+05,358183.000000,1.000000,NaN,5.412693e+12,1351.000000,NaN,NaN,NaN
75%,2.639780e+07,3.168058e+09,NaN,NaN,89725.000000,42200.000000,13802.400000,0.000000,0.0,0.0,...,42000.000000,5.474140e+05,632694.000000,1.000000,NaN,5.429261e+12,1466.250000,NaN,NaN,NaN
max,2.985042e+07,5.731247e+09,NaN,NaN,209400.000000,78089.890000,35600.000000,30284.430000,0.0,0.0,...,120000.000000,1.162220e+06,899914.000000,6.000000,NaN,5.489349e+12,1582.000000,NaN,NaN,NaN


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
ID,
25611457,6
25397560,4
26537906,4
26916796,4
26620130,4
...,...
20407121,1
20425943,1
20426447,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
NOMBRE CLIENTE,
Alvaro Camacho,9
Sandra Restrepo,8
Martha castano,8
Emelda Castro Hidalgo,8
ROCIO GOMEZ,7
...,...
Luz Cufiño,1
Elba Muñoz florez,1
Alvaro mera Mera benavides,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
TELÉFONO,
3012096436,9
3118691367,8
3103615550,8
3218614672,8
3014047657,8
...,...
3133006151,1
3102493528,1
3217779955,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
EMAIL,
jeidisaavedra74@gmail.com,2
neisermurillo771@gmail.com,2
hhh@gmail.com,2
yurenaalfa@hotmail.com,2
lolope10@hotmail.com,2
jose.edgar.trujillo.a@hotmail.com,1
prueba@ganual.com,1
martinezjimenezcarlosjulio@gmail.com,1
misia.pedro14@gmail.com,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
TIPO DE IDENTIFICACION,


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
NÚMERO GUIA,
014141495842,6
014141461740,4
014141496271,4
014141756952,4
014141587590,4
...,...
014139297162,1
014141330977,1
014141330909,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
ESTATUS,
ENTREGADO,592
CANCELADO,346
DEVOLUCION,66
RECHAZADO,8
INDEMNIZADA POR DROPI,5
DEVUELTA,3
RECOGIDA FALLIDA,2


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
TIPO DE ENVIO,
CON RECAUDO,1011
SIN RECAUDO,11


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
DEPARTAMENTO DESTINO,
CUNDINAMARCA,227
ANTIOQUIA,150
ATLANTICO,110
VALLE,78
SANTANDER,49
BOLIVAR,46
META,45
CORDOBA,40
MAGDALENA,29


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
CIUDAD DESTINO,
BOGOTA,178
BARRANQUILLA,80
MEDELLIN,74
CALI,38
CARTAGENA,34
...,...
SAN JUAN DE URABA,1
AGUACHICA,1
TORIBIO,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
DIRECCION,
"CL 3 # 14 C - 04 BR MIRAMAR, MIRAMAR",9
"CL 38 SUR # 43 - 40, ZONA CENTRO",8
Calle 28 # 22-69,8
"KR 52 B # 90 - 161 SANTA MONICA CRA 52 B 90161, SANTA MONICA",8
"KR 9 # 18 - 82 CENTRO PLAZA DE BOLIVAR LOCALCOMERCIAL, CENTRO HISTÓRICO",6
...,...
"AV CIRCUNVALAR 35 # 72 - 98 UB SERREZUELA AP 906 AP 906, EL TEJAR",1
"KR 24 C # 42 C SUR - 73 CS CLARET, CLARET",1
"CL 93 # 71 - 117 TO 2 AP 705 VILLA CAROLINA, CR. PASEO DEL PARQUE",1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
NOTAS,
Cepillo eléctrico 4 cabezales repuesto I: Negro ***,16
Cepillo eléctrico 4 cabezales repuesto I: Negro *** Cepillo eléctrico 4 cabezales repuesto I: Blanco ***,10
Cepillo Dientes Eléctrico Recargable: Rosado ***,10
Sábanas 180 hilos: Beige / Doble ***,9
Sábanas 180 hilos: Blanco / Doble ***,8
...,...
Sábanas 180 hilos: Rojo / King ***,1
"PROTECTOR MANIJA DROP: Toyota *** Kilómetro 2 variante cota condominio la vuelta al río 1 casa 19C,",1
Prueba,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
TRANSPORTADORA,
ENVIA,749
INTERRAPIDISIMO,158
SERVIENTREGA,38
99MINUTOS,28
COORDINADORA,24
TCC,21
VELOCES,4


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
SKU,
1323652951,74
null-37,48
'-1,44
'-3,43
'-19,41
...,...
null-31,1
ESSA0213-0,1
13,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
VARIACION ID,
862552.0,48
358163.0,41
358180.0,37
358179.0,31
358177.0,29
...,...
859765.0,1
252124.0,1
856091.0,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
PRODUCTO,
Sábanas Premium,589
Protector de colchón antifluido,138
Limpiador De Oidos Electrico Remove cera,74
Cepillo eléctrico 4 cabezales repuesto I,41
Cepillo de Dientes Recargable Black Days,24
Tensiometro De Brazo Digital Con Voz,23
Lonchera Térmica Para Almuerzo,21
Set Termo Temperatura Acero Inoxidable,21
CALCETINES DE ACUPUNTURA (TALLA UNICA),9


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
VARIACION,
Azul oscuro/Sencilla/,63
NEGRO,56
Beige/Doble/,54
Doble/Blanco/,42
Doble/Azul oscuro/,41
...,...
Vinotinto/Doble/,1
Semi doble/Azul oscuro/,1
MARRON CLARO,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
NOVEDAD,
COORDINAR LA ENTREGA,31
DIRECCION DESTINATARIO NO EXISTE,19
DESTINATARIO SE REHUSA A RECIBIR,17
DESTINATARIO FIJA FECHA Y HORA DE RECIBO MAYOR AL DIA SIGUIENTE,6
DESTINATARIO NO CANCELA C.O.D. LOGISTICA,3
DESTINATARIO SE TRASLADO,3
Rechaza Paquete,2
DIRECCION DESTINATARIO INCOMPLETA,2
NO LE INTERESA LA MERCANCÍA O CANCELA PEDIDO,2


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
FUE SOLUCIONADA LA NOVEDAD,
SI,51
NO,44


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
SOLUCIÓN,
DEVOLVER AL REMITENTE,20
"Se coordina entrega para martes 01 de octubre de 2024 en la direccion CL 73 # 39 - 40 APTO 201, cliente dejara dinero para mayor seguridad, se habla con el cliente al 3104529693",4
"Entregar martes 20 de agosto de 2024, el cliente va a dejar el dinero en la portería con el servicio de vigilancia",2
Devolución total del despacho - CARGADO A LA DIVISION 02 DEVOLVER AL REMITENTE,2
"El cliente confirma que la direccion es correcta Cra 103 # 6-85 Apto 111 Torre 2 Conjunto Ankara, Barrio Urbanización Ciudad Jardín en la ciudad de Cali. Llegando hay un letrero de MARVAL",2
"SE HABLA CON EL CLIENTE EN EL 3042578403, DICE QUE SI ESTA INTERESADO EN RECIBIR EL PAQUETE, SE COORDINA LA ENTREGA PARA MARTES 1 DE OCTUBRE",2
"Cliente informa direccion correcta KR 34 # 36 - 84, el cliente recomienda no ir entre las 12pm a 2pm porque es una oficina y a esa hora se encuentran fuera en hora de almuerzo",2
"Se coordina la entrega con el cliente para el dia de hoy 4 o mañana 5 de octubre, ambos días estará en el pueblo disponible para recibir el paquete, llamar al 3208022332",2
Se contacto al cliente señor Jose Edgar Trujillo Artunduaga CC 7701920 y confirma que el se acerca a reclamarlo en la oficina VILLAVICENCIO-META CALLE 27 # 26 - 03 BARRIO PORVENIR,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
OBSERVACIÓN,


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
ÚLTIMO MOVIMIENTO,
ENTREGADA DIGITALIZADA,454
DEVUELTA,47
ENTREGADA,32
ENTREGA VERIFICADA,23
CENTRO ACOPIO,21
ARCHIVADA,8
TRANSITO URBANO,6
TRANSITO NACIONAL,6
EN REPARTO,5


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
CONCEPTO ÚLTIMO MOVIMIENTO,
13,454
12,47
ENTREGADA,32
CENTRO ACOPIO,21
ARCHIVADA,8
TRANSITO URBANO,6
TRANSITO NACIONAL,6
6,5
DIGITALIZADA,4


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
UBICACIÓN DE ÚLTIMO MOVIMIENTO,
ESTADO,508
BOGOTA,5
CALI,5
BUCARAMANGA,4
CUCUTA,3
FLORENCIA,2
BARRANCABERMEJA,1
RIONEGRO (ANT),1
MEDELLIN,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
VENDEDOR,
Kelly Clavijo,4


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
TIPO DE TIENDA,
SHOPIFY,741
CHATCENTER,74


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
TIENDA,
Casa Práctica Market,815


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
ID DE ORDEN DE TIENDA,
5.407593e+12,4
5.416568e+12,4
5.415428e+12,4
5.411106e+12,4
5.434975e+12,4
...,...
5.374591e+12,1
5.405034e+12,1
5.401682e+12,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
NUMERO DE PEDIDO DE TIENDA,
1284.0,4
1384.0,4
1379.0,4
1329.0,4
1514.0,4
...,...
1212.0,1
1252.0,1
1244.0,1


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
TAGS,


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
CONTADOR DE INDEMNIZACIONES,


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
CONTADOR DE INDEMNIZACIONES,


'Pedidos y productos crudo: estadísticas de columnas categóricas'

,count
CONCEPTO ÚLTIMA INDENMIZACIÓN,


'Customers crudo: vista general de la tabla'

,ID,NAME,LASTNAME,PHONE,EMAIL,ADDRESS1,ADDRESS2,IDENTIFICATION
0,23302143,PILAR,VERGARA,3005750947,NaN,KR 58 # 79 - 128 AP 301,NaN,NaN
1,23771079,Maria Elena,De Lavalle,3178408257,NaN,CL 21 SUR # 41 - 117 UB CANTA PIEDRA AP 1017,NaN,NaN
2,25599918,Fabián,Quintero,3026083391,NaN,Carrera 54a # 122a-54 Recreo de los Frailes Ca...,NaN,NaN
3,23487936,Ledis María,Oviedo cordero,3106345535,NaN,Barrio Buenosaire Buenosaires 2106345535,NaN,NaN
4,20181023,Antonio. Carlos,Jaraba. Favila,3135467067,NaN,Carrera 5 # 1_35 Barrio. El. Machin Casa,NaN,NaN




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 544 entries, 0 to 543
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   ID              544 non-null    int64  
 1   NAME            544 non-null    object 
 2   LASTNAME        544 non-null    object 
 3   PHONE           544 non-null    int64  
 4   EMAIL           29 non-null     object 
 5   ADDRESS1        544 non-null    object 
 6   ADDRESS2        0 non-null      float64
 7   IDENTIFICATION  0 non-null      float64
dtypes: float64(2), int64(2), object(4)
memory usage: 34.1+ KB


'Customers crudo: información general del dataset'

None

'Customers crudo: estadísticas de columnas numéricas'

,ID,PHONE,ADDRESS2,IDENTIFICATION
count,5.440000e+02,5.440000e+02,0.0,0.0
mean,2.172256e+07,7.345258e+09,NaN,NaN
std,1.795142e+06,4.874509e+10,NaN,NaN
min,1.583868e+07,3.014048e+08,NaN,NaN
25%,2.017839e+07,3.106583e+09,NaN,NaN
50%,2.249818e+07,3.135079e+09,NaN,NaN
75%,2.304205e+07,3.178419e+09,NaN,NaN
max,2.559992e+07,5.732122e+11,NaN,NaN


'Customers crudo: estadísticas de columnas categóricas'

,count
ID,
23788150,1
23788406,1
23785328,1
23775314,1
23769398,1
...,...
20181023,1
23487936,1
25599918,1


'Customers crudo: estadísticas de columnas categóricas'

,count
NAME,
Martha,9
Jose,6
Carlos,6
Gloria,6
Sandra,5
...,...
Hernán Alexander,1
Carmen Sofis,1
Ivis del Carmen,1


'Customers crudo: estadísticas de columnas categóricas'

,count
LASTNAME,
Gomez,5
Rodríguez,5
Vargas,4
Leon,4
Lopez,3
...,...
Amaya,1
Ortiz,1
Peñuela,1


'Customers crudo: estadísticas de columnas categóricas'

,count
PHONE,
3148178661,2
3156403627,1
3107598865,1
3213766060,1
3173222446,1
...,...
3206418697,1
3233113432,1
3113100618,1


'Customers crudo: estadísticas de columnas categóricas'

,count
EMAIL,
yurenaalfa@hotmail.com,1
giraldovillaaugusto084@gmail.com,1
eliga122630@gmail.com,1
edol2008@hotmail.com,1
dani0479@hotmail.com,1
danayiret68@hotmail.com,1
lolope10@hotmail.com,1
daryrociorinconmorales@gmail.com,1
genesjohanys@gmail.com,1


'Customers crudo: estadísticas de columnas categóricas'

,count
ADDRESS1,
Calle 28 # 22-69,2
carrera 1A4ABIS # 76-34 barrio calimio norte casa,2
Calle 31 # 29-41 Apto 803 LA AURORA APTO803,1
Calle 34 6-38 Centro Lapista,1
Oficina Interrapidisimo Arauca,1
...,...
Cra 72 r bis n 42 b 80 sur Boita CRA72RBISN42B38SURAPTO302,1
Carrera 14 No 35-26 oficina 308 A Centro Oficina308A,1
Carrera 69 # 97 - 131 Castilla 503,1


'Customers crudo: estadísticas de columnas categóricas'

,count
ADDRESS2,


'Customers crudo: estadísticas de columnas categóricas'

,count
IDENTIFICATION,


'Marketing crudo: vista general de la tabla'

,Año,Mes,Canal,Costo
0,2024,Abril,Meta ads,116000.0
1,2024,Abril,Shopify,59347.0
2,2024,Mayo,Shopify,95980.0
3,2024,Mayo,Meta ads,420000.0
4,2024,Junio,Shopify,102343.0




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Año     22 non-null     int64  
 1   Mes     22 non-null     object 
 2   Canal   22 non-null     object 
 3   Costo   22 non-null     float64
dtypes: float64(1), int64(1), object(2)
memory usage: 836.0+ bytes


'Marketing crudo: información general del dataset'

None

'Marketing crudo: estadísticas de columnas numéricas'

,Año,Costo
count,22.0,2.200000e+01
mean,2024.0,6.581035e+05
std,0.0,9.005055e+05
min,2024.0,4.862278e+04
25%,2024.0,1.085492e+05
50%,2024.0,2.120944e+05
75%,2024.0,6.785540e+05
max,2024.0,3.250983e+06


'Marketing crudo: estadísticas de columnas categóricas'

,count
Año,
2024,22


'Marketing crudo: estadísticas de columnas categóricas'

,count
Mes,
Julio,3
Junio,3
Septiembre,3
Agosto,3
Octubre,3
Noviembre,3
Abril,2
Mayo,2


'Marketing crudo: estadísticas de columnas categóricas'

,count
Canal,
Meta ads,8
Shopify,8
Tiktok ads,6


# 🧹 Data Cleaning Plan – Documentación de limpieza de datos

A continuación se documentan las acciones de limpieza necesarias identificadas a partir del análisis exploratorio de los datasets:

---

# 📦 1. Limpieza: `ordenes_productos_raw`

## 🧱 1.1 Estructura y tipos de datos

- Convertir columnas de fecha y hora:
  - `FECHA` → datetime  
  - `HORA` → time  
  - `FECHA GUIA GENERADA` → datetime  
  - `FECHA DE NOVEDAD`, `FECHA DE SOLUCIÓN`, `FECHA DE ÚLTIMO MOVIMIENTO` → datetime  

- Convertir columnas numéricas correctamente:
  - `VARIACION ID`

---

## 🧽 1.2 Columnas completamente vacías o irrelevantes (eliminar)

Las siguientes columnas no contienen información útil:

- `TIPO DE IDENTIFICACION`  
- `NRO DE IDENTIFICACION`  
- `OBSERVACIÓN`  
- `TAGS`  
- `CONTADOR DE INDEMNIZACIONES`  
- `CONCEPTO ÚLTIMA INDENMIZACIÓN`  

---

## 🗑️ 1.3 Columnas sin valor analítico (eliminar)

Las siguientes columnas se eliminan por no aportar valor al análisis:

- `VENDEDOR`  
- `TIPO DE TIENDA`  
- `TIENDA`  
- `ID DE ORDEN DE TIENDA`  
- `NUMERO DE PEDIDO DE TIENDA`  

---

## ⚠️ 1.4 Columnas con alta cantidad de nulos

- `EMAIL` (muy incompleta)

👉 Acción:
- Mantener como atributo opcional para posibles análisis secundarios

---

## 🔁 1.5 Duplicados

- Existen IDs duplicados

⚠️ Regla crítica:
> NO eliminar duplicados directamente porque la tabla está a nivel **order-item (línea de producto)**

👉 Acción:
- Validar duplicados por combinación:
  - `ID` + `PRODUCTO` + `VARIACION`  
- Eliminar solo duplicados exactos

---

## 🧾 1.6 Estandarización de texto

Aplicar limpieza a columnas categóricas:

- Convertir a formato consistente (mayúsculas o título):
  - `NOMBRE CLIENTE`  
  - `CIUDAD DESTINO`  
  - `DEPARTAMENTO DESTINO`  
  - `PRODUCTO`  
  - `VARIACION`  

- Eliminar espacios innecesarios  
- Corregir caracteres especiales  

---

## 📊 1.7 Normalización de estados (`ESTATUS`)

Agrupar en categorías analíticas:

| ESTATUS ORIGINAL              | GRUPO            |
|-----------------------------|------------------|
| ENTREGADO                   | DELIVERED        |
| DEVOLUCION / DEVUELTA       | RETURNED         |
| CANCELADO / RECHAZADO       | CANCELLED        |
| RECOGIDA FALLIDA            | NOT_PROCESSED    |
| INDEMNIZADA POR DROPI       | CRITICAL_LOSS    |

👉 Crear nueva columna:
- `status_group`

---

## 📦 1.8 Validaciones numéricas

- Verificar:
  - `TOTAL DE LA ORDEN`  
  - `PRECIO PROVEEDOR`  
  - `PRECIO FLETE`  
  - `COSTO DEVOLUCION FLETE`  

👉 Detectar:
- Valores negativos  
- Valores en cero inconsistentes  

---

## 🧠 1.9 Feature Engineering (pre-modelado)

Renombrar columnas clave :

- `revenue` → TOTAL DE LA ORDEN  
- `cost_product` → PRECIO PROVEEDOR X CANTIDAD  
- `cost_shipping` → PRECIO FLETE  
- `cost_return` → COSTO DEVOLUCION FLETE  

Crear columnas Flags:
- `is_delivered`  
- `is_returned`  
- `is_cancelled`  
- `is_loss`  

---

## 🌍 1.10 Calidad geográfica

- Revisar:
  - `CIUDAD DESTINO`  
  - `DEPARTAMENTO DESTINO`  

👉 Estandarizar nombres (ej: Bogotá vs BOGOTA)

---

## 📞 1.11 Datos de cliente

- `TELÉFONO`:
  - Convertir a string  
  - Validar longitud (10 dígitos en Colombia)  

- `EMAIL`:
  - Mantener aunque incompleto  

---

---

# 👤 2. Limpieza: `clientes_raw`

## 🧱 2.1 Columnas a eliminar

- `ADDRESS2` (vacía)  
- `IDENTIFICATION` (vacía)  

---

## 🧽 2.2 Estandarización

- `NAME` y `LASTNAME`:
  - Unificar formato  
  - Eliminar caracteres extraños  

---

## 🔗 2.3 Creación de identificador

- Crear:
  - `full_name = NAME + LASTNAME`

👉 Útil para joins con órdenes

---

## 📞 2.4 Teléfono

- Convertir a string  
- Validar consistencia con órdenes  

---

## 📍 2.5 Dirección

- Limpiar `ADDRESS1`  
- Eliminar inconsistencias  

---

## 📧 2.6 Email

- Muy incompleto  
👉 Mantener como atributo opcional  

---

---

# 📣 3. Limpieza: `marketing_raw`

## 🧱 3.1 Tipos de datos

- `Año` → int (correcto)  
- `Mes` → categorizar / ordenar  
- `Costo` → float  

---

## 📅 3.2 Estandarización de fechas

- Crear columnas:
  - `month_number`  
  - `date` (YYYY-MM-01)  

👉 Necesario para joins con órdenes  

---

## 📊 3.3 Estandarización de canal

Valores detectados:
- Meta ads  
- Shopify  
- Tiktok ads  

👉 Acción:
- Normalizar nombres:
  - META_ADS  
  - TIKTOK_ADS  
  - SHOPIFY  

---

## ✅ 3.4 Validación de costos

- Se confirma que los valores de la columna `Costo` son correctos  
👉 No se requiere análisis de outliers  

---

---

# 🔗 4. Integración de datos (preparación)

## 🔄 Join strategy

- Orders ↔ Customers:
  - Por `TELÉFONO` o `NOMBRE`  

- Orders ↔ Marketing:
  - Por `FECHA` (mes/año)  

---

## ⚠️ Consideraciones clave

- Marketing está a nivel mensual  
- Orders está a nivel transaccional  

👉 Se requerirá:
- Prorrateo de costos de marketing  

---

# 🚨 5. Riesgos identificados

- Alta tasa de datos faltantes en clientes  
- Emails poco confiables  
- Variabilidad en direcciones  
- Diferencia de granularidad entre datasets  

---

# ✅ 6. Resultado esperado del cleaning

- Dataset limpio a nivel **order-item**  
- Estados normalizados  
- Costos correctamente definidos  
- Datos listos para:
  - Modelado  
  - Cálculo de KPIs  
  - Integración con marketing  

---

# Data Cleaning (Python - Pandas) 👇

In [3]:
# =========================
# 1. LIBRERÍAS
# =========================
import pandas as pd
import numpy as np

# =========================
# 2. COPIA DE SEGURIDAD
# =========================
orders = orders_products_raw.copy()
customers = customers_raw.copy()
marketing = marketing_raw.copy()

# =========================
# 3. LIMPIEZA: orders_products_raw
# =========================

# --- 3.1 Eliminar columnas irrelevantes o vacías ---
cols_drop_orders = [
    'TIPO DE IDENTIFICACION',
    'NRO DE IDENTIFICACION',
    'OBSERVACIÓN',
    'TAGS',
    'CONTADOR DE INDEMNIZACIONES',
    'CONCEPTO ÚLTIMA INDENMIZACIÓN',
    'VENDEDOR',
    'TIPO DE TIENDA',
    'TIENDA',
    'ID DE ORDEN DE TIENDA',
    'NUMERO DE PEDIDO DE TIENDA'
]

orders.drop(columns=cols_drop_orders, inplace=True, errors='ignore')

# --- 3.2 Conversión de fechas ---
date_cols = [
    'FECHA',
    'FECHA GUIA GENERADA',
    'FECHA DE NOVEDAD',
    'FECHA DE SOLUCIÓN',
    'FECHA DE ÚLTIMO MOVIMIENTO'
]

for col in date_cols:
    orders[col] = pd.to_datetime(orders[col], errors='coerce', dayfirst=True)

# --- 3.3 Limpieza de texto ---
text_cols = [
    'NOMBRE CLIENTE',
    'CIUDAD DESTINO',
    'DEPARTAMENTO DESTINO',
    'PRODUCTO',
    'VARIACION'
]

for col in text_cols:
    orders[col] = (
        orders[col]
        .astype(str)
        .str.strip()
        .str.upper()
    )

# --- 3.4 Teléfono como string ---
orders['TELÉFONO'] = orders['TELÉFONO'].astype(str).str.replace('.0', '', regex=False)

# --- 3.5 Normalización de ESTATUS ---
def map_status(status):
    if status in ['ENTREGADO']:
        return 'delivered'
    elif status in ['DEVOLUCION', 'DEVUELTA']:
        return 'returned'
    elif status in ['CANCELADO', 'RECHAZADO']:
        return 'cancelled'
    elif status in ['RECOGIDA FALLIDA']:
        return 'not_processed'
    elif status in ['INDEMNIZADA POR DROPI']:
        return 'critical_loss'
    else:
        return 'other'

orders['status_group'] = orders['ESTATUS'].apply(map_status)

# --- 3.6 Eliminar duplicados exactos (no por ID solamente) ---
orders.drop_duplicates(subset=['ID', 'PRODUCTO', 'VARIACION'], inplace=True)

# --- 3.7 Feature engineering (renombrado de columnas) ---

orders.rename(columns={
    'TOTAL DE LA ORDEN': 'revenue',
    'PRECIO PROVEEDOR X CANTIDAD': 'cost_product',
    'PRECIO FLETE': 'cost_shipping',
    'COSTO DEVOLUCION FLETE': 'cost_return'
}, inplace=True)

# estandarizar a snake_case
orders.columns = (
    orders.columns
    .str.strip()
    .str.lower()
    .str.replace(' ', '_')
    .str.replace('á', 'a')
    .str.replace('é', 'e')
    .str.replace('í', 'i')
    .str.replace('ó', 'o')
    .str.replace('ú', 'u')
    .str.replace('ñ', 'n')
)

# Flags
orders['is_delivered'] = (orders['status_group'] == 'delivered').astype(int)
orders['is_returned'] = (orders['status_group'] == 'returned').astype(int)
orders['is_cancelled'] = (orders['status_group'] == 'cancelled').astype(int)

# Definición de pérdida
orders['is_loss'] = np.where(
    orders['status_group'].isin(['returned', 'critical_loss']),
    1, 0
)

# --- 3.8 Validaciones básicas ---
# Reemplazar valores negativos por NaN (si existen)
num_cols = ['revenue', 'cost_product', 'cost_shipping', 'cost_return']

for col in num_cols:
    orders[col] = orders[col].apply(lambda x: np.nan if x < 0 else x)

# =========================
# 4. LIMPIEZA: customers_raw
# =========================

# --- 4.1 Eliminar columnas vacías ---
customers.drop(columns=['ADDRESS2', 'IDENTIFICATION'], inplace=True, errors='ignore')

# --- 4.2 Limpieza de nombres ---
customers['NAME'] = customers['NAME'].astype(str).str.strip().str.title()
customers['LASTNAME'] = customers['LASTNAME'].astype(str).str.strip().str.title()

# --- 4.3 Crear full_name ---
customers['full_name'] = customers['NAME'] + ' ' + customers['LASTNAME']

# --- 4.4 Teléfono como string ---
customers['PHONE'] = customers['PHONE'].astype(str).str.replace('.0', '', regex=False)

# --- 4.5 Limpieza de dirección ---
customers['ADDRESS1'] = customers['ADDRESS1'].astype(str).str.strip().str.upper()

# estandarizar a snake_case
customers.columns = (
    customers.columns
    .str.strip()
    .str.lower()
    .str.replace(' ', '_')
    .str.replace('á', 'a')
    .str.replace('é', 'e')
    .str.replace('í', 'i')
    .str.replace('ó', 'o')
    .str.replace('ú', 'u')
    .str.replace('ñ', 'n')
)

# =========================
# 5. LIMPIEZA: marketing_raw
# =========================

# --- 5.1 Estandarizar nombres ---

marketing.columns = (
    marketing.columns
    .str.strip()
    .str.lower()
    .str.replace('ñ', 'n')
)
marketing.rename(columns={
    'canal': 'channel',
    'costo': 'marketing_cost'
    }, inplace=True)

marketing['channel'] = marketing['channel'].replace({
    'Meta ads': 'meta_ads',
    'Tiktok ads': 'tiktok_ads',
    'Shopify': 'shopify'
})

# --- 5.2 Crear número de mes ---
month_map = {
    'enero': 1, 'febrero': 2, 'marzo': 3, 'abril': 4,
    'mayo': 5, 'junio': 6, 'julio': 7, 'agosto': 8,
    'septiembre': 9, 'octubre': 10, 'noviembre': 11, 'diciembre': 12
}
marketing['mes'] = (marketing['mes'].astype(str).str.strip().str.lower())
marketing['month_number'] = marketing['mes'].map(month_map)

# --- 5.3 Crear fecha ---
marketing['date'] = pd.to_datetime(
    dict(year=marketing['ano'], month=marketing['month_number'], day=1)
)
marketing['marketing_cost'] = pd.to_numeric(marketing['marketing_cost'], errors='coerce')

# =========================
# 6. OUTPUT LIMPIO
# =========================

orders_clean = orders.copy()
customers_clean = customers.copy()
marketing_clean = marketing.copy()

# Vista rápida
print("Orders limpio:", orders_clean.shape)
print("Customers limpio:", customers_clean.shape)
print("Marketing limpio:", marketing_clean.shape)

Orders limpio: (1022, 46)
Customers limpio: (544, 7)
Marketing limpio: (22, 6)


In [4]:
# Inspeccion de datasets despues de limpieza (Python)

display("Ordenes y pedidos limpio: información general del dataset", orders_clean.info())
print(f"\n")
display("Clientes limpio: información general del dataset", customers_clean.info())
print(f"\n")
display("Marketing limpio: información general del dataset", marketing_clean.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1022 entries, 0 to 1021
Data columns (total 46 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   id                              1022 non-null   int64         
 1   hora                            1022 non-null   object        
 2   fecha                           1022 non-null   datetime64[ns]
 3   nombre_cliente                  1022 non-null   object        
 4   telefono                        1022 non-null   object        
 5   email                           33 non-null     object        
 6   numero_guia                     676 non-null    object        
 7   estatus                         1022 non-null   object        
 8   tipo_de_envio                   1022 non-null   object        
 9   departamento_destino            1022 non-null   object        
 10  ciudad_destino                  1022 non-null   object        
 11  dire

'Ordenes y pedidos limpio: información general del dataset'

None



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 544 entries, 0 to 543
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   id         544 non-null    int64 
 1   name       544 non-null    object
 2   lastname   544 non-null    object
 3   phone      544 non-null    object
 4   email      29 non-null     object
 5   address1   544 non-null    object
 6   full_name  544 non-null    object
dtypes: int64(1), object(6)
memory usage: 29.9+ KB


'Clientes limpio: información general del dataset'

None



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   ano             22 non-null     int64         
 1   mes             22 non-null     object        
 2   channel         22 non-null     object        
 3   marketing_cost  22 non-null     float64       
 4   month_number    22 non-null     int64         
 5   date            22 non-null     datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(2), object(2)
memory usage: 1.2+ KB


'Marketing limpio: información general del dataset'

None

# Data Modeling (SQL - enfoque práctico) 👇

---

## 📌 1. Enfoque del modelo

Para este proyecto se implementa un **modelo denormalizado (wide table)** optimizado para análisis en herramientas de BI como Tableau.

A diferencia de un modelo estrella tradicional, se prioriza:

- Simplicidad en el análisis  
- Mejor performance en visualización  
- Eliminación de joins en la capa de BI  

---

### 🎯 Objetivo del modelo

Construir una tabla analítica única que permita:

- Analizar rentabilidad a nivel de pedido y producto  
- Integrar operación + marketing  
- Facilitar el cálculo de KPIs en Tableau  
- Reducir complejidad del modelo  

---

## ⭐ 2. Tabla principal

### `fact_final`

Granularidad:
> Una fila por producto dentro de cada pedido  

Esta tabla consolida toda la información necesaria para análisis:

- Datos de pedidos (orders)  
- Información de clientes (vía teléfono)  
- Costos operativos  
- Costos de marketing (prorrateados)  

---

## 📊 3. Estructura de la tabla

### 📅 Dimensiones
- fecha  
- year  
- month  
- month_number  
- product  
- city  
- department  
- status_group  
- channel  

---

### 💰 Métricas base
- revenue  
- cost_product  
- cost_shipping  
- cost_return  
- marketing_cost  

---

### 📈 Métricas calculadas
- profit_operativo  
- profit_neto  

---

### 🚩 Flags operativos
- is_delivered  
- is_returned  
- is_cancelled  
- is_loss  

---

## ⚠️ 4. Consideración clave: Marketing

El dataset de marketing se encuentra a nivel **mensual**, mientras que los pedidos están a nivel **transaccional**.

Por lo tanto, es necesario distribuir el costo de marketing entre los pedidos.

---

### 📣 Prorrateo de marketing

Se utiliza una distribución uniforme:

marketing_cost_per_order = costo_marketing_mensual / total_pedidos_mes


Este valor se asigna a cada pedido dentro del mes correspondiente:

- `marketing_cost`

---

### 🎯 Implicación

Permite calcular:

- Profit real (incluyendo marketing)  
- Costo por pedido  
- Costo por pedido entregado  

---

## 🚀 5. Resultado del modelo

Se obtiene una tabla optimizada para:

- Consultas SQL analíticas  
- Integración directa con Tableau  
- Cálculo dinámico de KPIs  
- Análisis end-to-end del negocio  

---

## 🧠 Nota de diseño

Inicialmente se consideró un modelo tipo Star Schema (tablas de dimensiones), pero para efectos prácticos del análisis y visualización se optó por un modelo **denormalizado**, más eficiente para este caso de uso.

### Preparación (Python + SQLite)

In [5]:
import sqlite3
import pandas as pd

# Crear conexión
conn = sqlite3.connect("ecommerce.db")

# Cargar tablas limpias
orders_clean.to_sql("orders_clean", conn, if_exists="replace", index=False)
customers_clean.to_sql("customers_clean", conn, if_exists="replace", index=False)
marketing_clean.to_sql("marketing_clean", conn, if_exists="replace", index=False)

# Helper para ejecutar SQL y ver resultados
def run_query(query):
    return pd.read_sql_query(query, conn)

# Data Integration (tabla final unificada) 👇

---

## 📌 1. Fuentes de datos

- `orders_clean` → información transaccional y costos operativos  
- `customers_clean` → información de clientes  
- `marketing_clean` → inversión en adquisición de clientes  

---

## 🧱 2. Resultado

Se construye la tabla:  `fact_final`

Esta tabla contiene:

- Información transaccional integrada  
- Costos operativos ajustados  
- Costo de marketing prorrateado  


### Crear tabla `fact_final`

In [6]:
conn.execute("DROP TABLE IF EXISTS fact_final")

query = """
CREATE TABLE fact_final AS
WITH marketing_monthly AS (
    SELECT
        STRFTIME('%Y', date) AS year,
        STRFTIME('%m', date) AS month,
        SUM(marketing_cost) AS total_marketing_cost
    FROM marketing_clean
    GROUP BY year, month
),
orders_per_month AS (
    SELECT
        STRFTIME('%Y', fecha) AS year,
        STRFTIME('%m', fecha) AS month,
        COUNT(DISTINCT id) AS total_orders
    FROM orders_clean
    GROUP BY year, month
)

SELECT
    -- Identificadores
    o.id AS order_id,
    DATE(o.fecha) AS date,

    -- Cliente
    c.id AS customer_id,

    -- Producto
    o.producto AS product_name,
    o.variacion AS variation,
    o.cantidad AS quantity,

    -- Ubicación
    o.departamento_destino AS department,
    o.ciudad_destino AS city,

    -- Tiempo
    CAST(STRFTIME('%Y', o.fecha) AS INTEGER) AS year,
    CAST(STRFTIME('%m', o.fecha) AS INTEGER) AS month,

    -- Estado
    o.status_group,

    -- Revenue real
    CASE
        WHEN o.status_group = 'delivered' THEN o.revenue
        ELSE 0
    END AS revenue_real,

    -- Costos reales
    CASE
        WHEN o.status_group IN ('delivered','returned','critical_loss')
        THEN o.cost_product ELSE 0
    END AS cost_product_real,

    CASE
        WHEN o.status_group IN ('delivered','returned','critical_loss')
        THEN o.cost_shipping ELSE 0
    END AS cost_shipping_real,

    CASE
        WHEN o.status_group = 'returned'
        THEN o.cost_return ELSE 0
    END AS cost_return_real,

    -- Marketing prorrateado (unitario)
    (m.total_marketing_cost * 1.0 / opm.total_orders) AS marketing_cost_per_order,

       -- Flags
    o.is_delivered,
    o.is_returned,
    o.is_cancelled,
    o.is_loss

FROM orders_clean o

LEFT JOIN customers_clean c
    ON o.telefono = c.phone

LEFT JOIN marketing_monthly m
    ON STRFTIME('%Y', o.fecha) = m.year
    AND STRFTIME('%m', o.fecha) = m.month

LEFT JOIN orders_per_month opm
    ON STRFTIME('%Y', o.fecha) = opm.year
    AND STRFTIME('%m', o.fecha) = opm.month;
"""
conn.execute(query)

### Validación

In [7]:
run_query("""
SELECT
    COUNT(*) AS total_rows,
    COUNT(DISTINCT order_id) AS total_orders
FROM fact_final;
""")

,total_rows,total_orders
0,1024,645


# Exploratory Data Analysis (EDA) 👇

## 📌 1. Enfoque del análisis
En esta etapa se realiza un análisis exploratorio sobre la tabla integrada `fact_final`, con el objetivo de entender el comportamiento general del negocio bajo el modelo COD.

El EDA permite:

- Comprender la distribución de los pedidos
- Identificar patrones operativos
- Analizar cómo se comportan ingresos y costos a nivel descriptivo
- Detectar posibles anomalías en los datos

---

## 🎯 2. Objetivos del EDA

- Analizar la distribución de pedidos por `status_group`
- Evaluar la proporción de pedidos entregados vs no efectivos
- Explorar el comportamiento de `revenue` y costos
- Analizar la frecuencia de pérdidas operativas (`is_loss`)
- Entender la variabilidad de costos logísticos

---

## 🔍 3. Análisis desarrollados

### 📦 Distribución de pedidos
- Conteo total de registros
- Número de órdenes únicas
- Distribución por estado:
  - delivered
  - returned
  - cancelled
  - not_processed
  - critical_loss

---

### 💰 Análisis de revenue
- Distribución de `revenue`
- Comparación descriptiva entre pedidos entregados vs no entregados
- Identificación de valores atípicos

---

### 💸 Análisis de costos
Se analiza la distribución de:

- `cost_product`
- `cost_shipping`
- `cost_return`

Se evalúa:

- Dispersión de los costos
- Presencia de valores altos
- Diferencias entre tipos de pedidos

---

### ⚠️ Análisis de pérdidas
- Proporción de pedidos con `is_loss = 1`
- Frecuencia de pérdidas operativas
- Relación entre estados (`returned`, `critical_loss`) y pérdidas

---

### 📅 Análisis temporal
- Volumen de pedidos por mes
- Evolución de estados de pedido en el tiempo
- Comportamiento operativo básico
- Promedio de costos de marketing por mes

---

## 🧠 4. Enfoque metodológico

El análisis se realiza mediante:

- Consultas SQL
- Agregaciones simples
- Conteos y proporciones
- Estadísticas descriptivas

---

## ⚠️ 5. Consideraciones

- Este análisis es **exploratorio (no conclusivo)**
- No se calculan KPIs finales en esta etapa
- No se incluyen métricas de marketing ni rentabilidad neta
- No se generan recomendaciones aún

---

## 🚀 6. Resultado esperado

Un entendimiento claro de:

- Cómo se distribuyen los pedidos en el negocio
- Qué proporción de pedidos es efectiva vs fallida
- Cómo se comportan ingresos y costos
- Qué tan frecuentes son las pérdidas operativas

Este análisis sirve como base para el siguiente paso: **KPI Calculation**

### Proporción de pedidos (participación %)

In [8]:
query = """
SELECT
    status_group,
    COUNT(*) AS total_orders,
    ROUND(100.0 * COUNT(*) / SUM(COUNT(*)) OVER (), 2) AS percentage
FROM fact_final
GROUP BY status_group
ORDER BY total_orders DESC;
"""
run_query(query)

,status_group,total_orders,percentage
0,delivered,592,57.81
1,cancelled,355,34.67
2,returned,70,6.84
3,critical_loss,5,0.49
4,not_processed,2,0.20


### Distribución de revenue

In [9]:
query = """
SELECT
    MIN(revenue_real) AS min_revenue,
    MAX(revenue_real) AS max_revenue,
    AVG(revenue_real) AS avg_revenue,
    SUM(revenue_real) AS total_revenue
FROM fact_final;
"""
run_query(query)

,min_revenue,max_revenue,avg_revenue,total_revenue
0,0,209400.0,46089.157246,47195297.02


### Revenue por tipo de pedido

In [10]:
query = """
SELECT
    status_group,
    SUM(revenue_real) AS total_revenue,
    AVG(revenue_real) AS avg_revenue
FROM fact_final
GROUP BY status_group
ORDER BY total_revenue DESC;
"""
run_query(query)

,status_group,total_revenue,avg_revenue
0,delivered,47195297.02,79721.785507
1,returned,0.00,0.000000
2,not_processed,0.00,0.000000
3,critical_loss,0.00,0.000000
4,cancelled,0.00,0.000000


### Distribución de costos

In [11]:
query = """
SELECT
    AVG(cost_product_real) AS avg_product_cost,
    AVG(cost_shipping_real) AS avg_shipping_cost,
    AVG(cost_return_real) AS avg_return_cost
FROM fact_final;
"""
run_query(query)

,avg_product_cost,avg_shipping_cost,avg_return_cost
0,23256.533203,6833.629639,953.526299


### Costos por tipo de pedido

In [12]:
query = """
SELECT
    status_group,
    AVG(cost_product_real) AS avg_product_cost,
    AVG(cost_shipping_real) AS avg_shipping_cost,
    AVG(cost_return_real) AS avg_return_cost
FROM fact_final
GROUP BY status_group;
"""
run_query(query)

,status_group,avg_product_cost,avg_shipping_cost,avg_return_cost
0,cancelled,0.000000,0.000000,0.000000
1,critical_loss,35200.000000,14260.000000,0.000000
2,delivered,36466.199324,10263.323497,0.000000
3,not_processed,0.000000,0.000000,0.000000
4,returned,29295.714286,12149.274857,13948.727571


### Análisis de pérdidas

In [13]:
query = """
SELECT
    is_loss,
    COUNT(*) AS total_orders
FROM fact_final
GROUP BY is_loss;
"""
run_query(query)

,is_loss,total_orders
0,0,949
1,1,75


### Pérdidas por tipo de estado

In [14]:
query = """
SELECT
    status_group,
    SUM(is_loss) AS loss_orders,
    COUNT(*) AS total_orders
FROM fact_final
GROUP BY status_group
ORDER BY loss_orders DESC;
"""
run_query(query)

,status_group,loss_orders,total_orders
0,returned,70,70
1,critical_loss,5,5
2,not_processed,0,2
3,delivered,0,592
4,cancelled,0,355


### Evolución mensual de pedidos

In [15]:
query = """
SELECT
    year,
    month,
    COUNT(*) AS total_orders
FROM fact_final
GROUP BY year, month
ORDER BY year, month;
"""
run_query(query)

,year,month,total_orders
0,2024,4,3
1,2024,5,10
2,2024,6,20
3,2024,7,115
4,2024,8,90
5,2024,9,470
6,2024,10,314
7,2024,11,2


### Evolución por estado

In [16]:
query = """
SELECT
    year,
    month,
    status_group,
    COUNT(*) AS total_orders
FROM fact_final
GROUP BY year, month, status_group
ORDER BY year, month;
"""
run_query(query)

,year,month,status_group,total_orders
0,2024,4,cancelled,1
1,2024,4,delivered,2
2,2024,5,cancelled,2
3,2024,5,critical_loss,4
4,2024,5,delivered,2
5,2024,5,returned,2
6,2024,6,cancelled,6
7,2024,6,critical_loss,1
8,2024,6,delivered,13
9,2024,7,cancelled,21


### Distribución de marketing por mes

In [17]:
query = """
SELECT
    year,
    month,
    AVG(marketing_cost_per_order) AS avg_marketing_cost
FROM fact_final
GROUP BY year, month
ORDER BY year, month;
"""
run_query(query)

,year,month,avg_marketing_cost
0,2024,4,87673.500000
1,2024,5,51598.000000
2,2024,6,40046.316447
3,2024,7,26058.172321
4,2024,8,26374.532895
5,2024,9,12638.233083
6,2024,10,27178.763500
7,2024,11,236405.785400


### Validación rápida de consistencia

In [18]:
query = """
SELECT
    COUNT(*) AS total_rows,
    COUNT(DISTINCT order_id) AS total_orders,
    SUM(is_delivered) AS delivered_orders,
    SUM(is_returned) AS returned_orders,
    SUM(is_cancelled) AS cancelled_orders
FROM fact_final;
"""
run_query(query)

,total_rows,total_orders,delivered_orders,returned_orders,cancelled_orders
0,1024,645,592,70,355


# KPI Calculation (Operativos + Marketing) 👇

## 📌 1. Enfoque del análisis
En esta etapa se calculan los indicadores clave de desempeño (KPIs) a partir de la tabla integrada `fact_final`.

A diferencia del EDA, aquí se construyen métricas agregadas que permiten evaluar de forma clara:

- La rentabilidad del negocio
- La eficiencia operativa
- El impacto del marketing

Todos los KPIs se calculan utilizando consultas SQL sobre datos ya integrados y limpios.

---

## 🎯 2. Objetivos de los KPIs

- Medir la rentabilidad del negocio (operativa y neta)
- Evaluar el desempeño logístico
- Cuantificar el impacto de devoluciones y cancelaciones
- Incorporar el costo de marketing en el análisis
- Identificar la magnitud de las pérdidas operativas

---

## 💰 3. KPIs Financieros

Se calculan métricas agregadas a nivel global:

### Ingresos y costos
- **Revenue Total** → suma de `revenue`
- **Costo de producto (COGS)** → suma de `cost_product`
- **Costo de envíos** → suma de `cost_shipping`
- **Costo de devoluciones** → suma de `cost_return`
- **Costo de marketing total** → suma de `marketing_cost`

---

### 📈 Rentabilidad

- **Profit Operativo**
  
  `revenue - (cost_product + cost_shipping + cost_return)`

- **Profit Neto**
  
  `profit_operativo - marketing_cost`

- **Margen Operativo**
  
  `profit_operativo / revenue`

- **Margen Neto**
  
  `profit_neto / revenue`

---

## 🚚 4. KPIs Operativos

- **Fulfillment Rate**
  
  `% pedidos entregados` → `is_delivered`

- **Return Rate**
  
  `% pedidos devueltos` → `is_returned`

- **Cancellation Rate**
  
  `% pedidos cancelados` → `is_cancelled`

- **Non-Processed Rate**
  
  `% pedidos no procesados` → `status_group = 'not_processed'`

---

## ⚠️ 5. KPIs de Riesgo

- **Loss Rate**
  
  `% pedidos con pérdida` → `is_loss`

- **Critical Loss Rate**
  
  `% pedidos con pérdida crítica` → `status_group = 'critical_loss'`

---

## 📣 6. KPIs de Marketing

- **Marketing Cost Total**
  
  suma de `marketing_cost`

- **Marketing Cost per Order**
  
  `marketing_cost / total_orders`

- **Marketing Cost per Delivered Order**
  
  `marketing_cost / delivered_orders`

---

## 🧠 7. Consideraciones metodológicas

- Los KPIs se calculan a nivel agregado (no por producto)
- Se utilizan métricas ya definidas en `fact_final`
- El costo de marketing ya está prorrateado por pedido
- Se respeta la lógica del modelo COD:
  - No todos los pedidos generan ingresos
  - Algunos pedidos generan costos sin revenue
- Los flags (`is_delivered`, `is_returned`, etc.) simplifican los cálculos

---

## ⚠️ 8. Consideraciones importantes

- No se realizan segmentaciones avanzadas en este paso
- No se incluyen visualizaciones
- No se generan insights ni recomendaciones aún
- Este paso es exclusivamente de cálculo de métricas

---

## 🚀 9. Resultado esperado

Un conjunto claro de indicadores que permiten responder:

- ¿El negocio es rentable?
- ¿Cuál es el impacto real del marketing?
- ¿Qué proporción de pedidos genera pérdidas?
- ¿Qué tan eficiente es la operación?

Estos KPIs serán utilizados directamente en el **dashboard en Tableau**.

### Base a nivel pedido

In [19]:
query = """
WITH order_level AS (
    SELECT
        order_id,

        SUM(revenue_real) AS revenue,
        SUM(cost_product_real) AS cost_product,
        SUM(cost_shipping_real) AS cost_shipping,
        SUM(cost_return_real) AS cost_return,

        MAX(marketing_cost_per_order) AS marketing_cost,

        MAX(is_delivered) AS is_delivered,
        MAX(is_returned) AS is_returned,
        MAX(is_cancelled) AS is_cancelled,
        MAX(is_loss) AS is_loss,

        MAX(CASE WHEN status_group = 'critical_loss' THEN 1 ELSE 0 END) AS is_critical_loss

    FROM fact_final
    GROUP BY order_id
)
SELECT * FROM order_level LIMIT 10;
"""
run_query(query)

,order_id,revenue,cost_product,cost_shipping,cost_return,marketing_cost,is_delivered,is_returned,is_cancelled,is_loss,is_critical_loss
0,17719911,0.0,0,0.0,0.0,87673.5,0,0,1,0,0
1,17991863,105000.0,46500,15875.0,0.0,87673.5,1,0,0,0,0
2,18250379,0.0,40000,15700.0,0.0,51598.0,0,0,0,1,1
3,18270877,0.0,40000,15700.0,0.0,51598.0,0,0,0,1,1
4,18275377,0.0,0,0.0,0.0,51598.0,0,0,1,0,0
5,18323622,0.0,40000,15700.0,0.0,51598.0,0,0,0,1,1
6,18450364,0.0,40000,12100.0,0.0,51598.0,0,0,0,1,1
7,18571749,0.0,0,0.0,0.0,51598.0,0,0,1,0,0
8,18572984,165900.0,95000,18775.0,0.0,51598.0,1,0,0,0,0
9,18581615,0.0,95000,14406.5,24006.5,51598.0,0,1,0,1,0


### KPIs financieros principales

In [20]:
query = """
WITH order_level AS (
    SELECT
        order_id,
        SUM(revenue_real) AS revenue,
        SUM(cost_product_real) AS cost_product,
        SUM(cost_shipping_real) AS cost_shipping,
        SUM(cost_return_real) AS cost_return,
        MAX(marketing_cost_per_order) AS marketing_cost
    FROM fact_final
    GROUP BY order_id
)

SELECT
    SUM(revenue) AS total_revenue,
    SUM(cost_product) AS total_cogs,
    SUM(cost_shipping) AS total_shipping,
    SUM(cost_return) AS total_return_cost,
    SUM(marketing_cost) AS total_marketing,

    -- Profit operativo
    SUM(revenue - cost_product - cost_shipping - cost_return) AS profit_operativo,

    -- Profit neto
    SUM(revenue - cost_product - cost_shipping - cost_return - marketing_cost) AS profit_neto

FROM order_level;
"""
run_query(query)


,total_revenue,total_cogs,total_shipping,total_return_cost,total_marketing,profit_operativo,profit_neto
0,47195297.02,23814690,6997636.75,976410.93,1.447828e+07,15406559.34,928281.4296


### Márgenes

In [21]:
query = """
WITH order_level AS (
    SELECT
        order_id,
        SUM(revenue_real) AS revenue,
        SUM(cost_product_real) AS cost_product,
        SUM(cost_shipping_real) AS cost_shipping,
        SUM(cost_return_real) AS cost_return,
        MAX(marketing_cost_per_order) AS marketing_cost
    FROM fact_final
    GROUP BY order_id
)

SELECT
    SUM(revenue) AS total_revenue,

    -- Margen operativo
    ROUND(
        SUM(revenue - cost_product - cost_shipping - cost_return) * 1.0
        / NULLIF(SUM(revenue),0), 4
    ) AS margin_operativo,

    -- Margen neto
    ROUND(
        SUM(revenue - cost_product - cost_shipping - cost_return - marketing_cost) * 1.0
        / NULLIF(SUM(revenue),0), 4
    ) AS margin_neto

FROM order_level;
"""
run_query(query)

,total_revenue,margin_operativo,margin_neto
0,47195297.02,0.3264,0.0197


### KPIs operativos

In [22]:
query = """
WITH order_level AS (
    SELECT
        order_id,
        MAX(is_delivered) AS is_delivered,
        MAX(is_returned) AS is_returned,
        MAX(is_cancelled) AS is_cancelled,
        MAX(CASE WHEN status_group = 'not_processed' THEN 1 ELSE 0 END) AS is_not_processed
    FROM fact_final
    GROUP BY order_id
)

SELECT
    COUNT(*) AS total_orders,

    ROUND(SUM(is_delivered) * 1.0 / COUNT(*), 4) AS fulfillment_rate,
    ROUND(SUM(is_returned) * 1.0 / COUNT(*), 4) AS return_rate,
    ROUND(SUM(is_cancelled) * 1.0 / COUNT(*), 4) AS cancellation_rate,
    ROUND(SUM(is_not_processed) * 1.0 / COUNT(*), 4) AS not_processed_rate

FROM order_level;
"""
run_query(query)

,total_orders,fulfillment_rate,return_rate,cancellation_rate,not_processed_rate
0,645,0.5798,0.0853,0.3256,0.0016


### KPIs de riesgo

In [23]:
query = """
WITH order_level AS (
    SELECT
        order_id,
        MAX(is_loss) AS is_loss,
        MAX(CASE WHEN status_group = 'critical_loss' THEN 1 ELSE 0 END) AS is_critical_loss
    FROM fact_final
    GROUP BY order_id
)

SELECT
    COUNT(*) AS total_orders,

    ROUND(SUM(is_loss) * 1.0 / COUNT(*), 4) AS loss_rate,
    ROUND(SUM(is_critical_loss) * 1.0 / COUNT(*), 4) AS critical_loss_rate

FROM order_level;
"""
run_query(query)

,total_orders,loss_rate,critical_loss_rate
0,645,0.093,0.0078


### KPIs de marketing

In [24]:
query = """
WITH order_level AS (
    SELECT
        order_id,
        MAX(marketing_cost_per_order) AS marketing_cost,
        MAX(is_delivered) AS is_delivered
    FROM fact_final
    GROUP BY order_id
)

SELECT
    SUM(marketing_cost) AS total_marketing_cost,

    AVG(marketing_cost) AS marketing_cost_per_order,

    SUM(marketing_cost) * 1.0 / NULLIF(SUM(is_delivered),0)
        AS marketing_cost_per_delivered_order

FROM order_level;
"""
run_query(query)

,total_marketing_cost,marketing_cost_per_order,marketing_cost_per_delivered_order
0,1.447828e+07,22446.942497,38711.973022


### Profit por tipo de pedido

In [25]:
query = """
WITH order_level AS (
    SELECT
        order_id,
        MAX(status_group) AS status_group,

        SUM(revenue_real) AS revenue,
        SUM(cost_product_real) AS cost_product,
        SUM(cost_shipping_real) AS cost_shipping,
        SUM(cost_return_real) AS cost_return,
        MAX(marketing_cost_per_order) AS marketing_cost
    FROM fact_final
    GROUP BY order_id
)

SELECT
    status_group,
    COUNT(*) AS total_orders,

    SUM(revenue - cost_product - cost_shipping - cost_return) AS profit_operativo,
    SUM(revenue - cost_product - cost_shipping - cost_return - marketing_cost) AS profit_neto

FROM order_level
GROUP BY status_group
ORDER BY profit_neto DESC;
"""
run_query(query)

,status_group,total_orders,profit_operativo,profit_neto
0,delivered,374,19531419.51,1.106143e+07
1,not_processed,1,0.00,-1.263823e+04
2,critical_loss,5,-247300.00,-4.937383e+05
3,cancelled,210,0.00,-4.450910e+06
4,returned,55,-3877560.17,-5.175859e+06


# Dashboard: Dropshipping Profitability Analytics 👇  
End-to-End Profit Optimization (Operación + Marketing) – Modelo COD

---

## 🎯 1. Objetivo del dashboard

Este dashboard está diseñado para evaluar la rentabilidad real del negocio utilizando la tabla integrada `fact_final`.

Integra:

- Ingresos (revenue)
- Costos de producto
- Costos logísticos
- Costos de marketing (prorrateados)

Permite responder:

- ¿El negocio es rentable después de marketing?
- ¿Dónde se generan las pérdidas?
- ¿Qué tan eficiente es la operación?
- ¿Cómo impacta el marketing en el resultado final?

---

## 🧱 2. Estructura del dashboard

El dashboard se organiza en 4 vistas principales:

---

### 🟦 1. Executive Overview

📌 **Objetivo**  
Visión general del desempeño del negocio

🔢 **KPIs principales**
- Revenue Total
- Cost Product Total
- Cost Shipping Total
- Cost Return Total
- Marketing Cost Total
- Profit Operativo (calculado en Tableau)
- Profit Neto (calculado en Tableau)
- Margen Neto (%)
- Loss Rate (%)

📊 **Visualizaciones**
- Revenue vs Costos (barra apilada)
- Distribución de pedidos por `status_group` (donut)
- Evolución de pedidos por fecha (línea)

🎛️ **Filtros**
- Fecha
- Producto
- Ciudad
- Departamento

---

### 🟩 2. Unit Economics

📌 **Objetivo**  
Entender la economía por pedido

🔢 **KPIs**
- AOV (Average Order Value)
- Cost per Order
- Marketing Cost per Order
- Profit por pedido (calculado en Tableau)

📊 **Visualizaciones**
- Descomposición de costos por pedido (barra)
  - Revenue per Order
  - Cost Product per Order
  - Cost Shipping per Order
  - Cost Return per Order
  - Marketing Cost per Order
- Comparación por tipo de pedido (`status_group`)

---

### 🟥 3. Operational Performance

📌 **Objetivo**  
Evaluar eficiencia logística

🔢 **KPIs**
- Fulfillment Rate
- Return Rate
- Cancellation Rate
- Loss Rate

📊 **Visualizaciones**
- Distribución de pedidos por estado
- Evolución de estados en el tiempo
- Costos logísticos:
  - Shipping Cost
  - Return Cost

---

### 🟪 4. Marketing Performance

📌 **Objetivo**  
Evaluar impacto del marketing

🔢 **KPIs**
- Total Marketing Cost
- Marketing Cost per Order
- Marketing Cost per Delivered Order
- ROAS (calculado en Tableau)

📊 **Visualizaciones**
- Costo por canal (barra)
- Revenue vs Marketing Cost

---

## 🧠 3. Insights esperados

El dashboard permitirá identificar:

- Impacto del marketing en la rentabilidad
- Proporción de pedidos con pérdida
- Efecto de devoluciones en costos
- Eficiencia operativa del modelo COD
- Relación entre volumen de pedidos y resultados

---

## 🏗️ 4. Fuente de datos

### Tabla principal:
`fact_final`

📌 **Campos utilizados**
- date
- channel
- product
- city
- department
- status_group
- revenue
- cost_product
- cost_shipping
- cost_return
- marketing_cost
- is_delivered
- is_returned
- is_cancelled
- is_loss

---

## 🚀 5. Resultado esperado

Un dashboard que permita:

- Evaluar la rentabilidad real del negocio
- Entender el impacto del modelo COD
- Identificar pérdidas operativas
- Analizar el impacto del marketing
- Servir como base para toma de decisiones

### **📦 Exportación**: Una vez finalizado todo el análisis se exporta la tabla `fact_final` con la que se desarrollará el dashboard.

In [26]:
# exportar dataset fact_final

fact_final = run_query("SELECT * FROM fact_final")
fact_final.to_csv('fact_final.csv', index=False)

### 🔗 En el siguiente link puedes ver el dashboard en Tableau Public:

 - https://public.tableau.com/app/profile/giovanni.ramirez1825/viz/Anlisisdecomercioelectrnicoydropshipping/ExecutiveOverview?publish=yes

# Insights & Business Recommendations 👇


## 🟦 1. Insight Principal: Rentabilidad Baja pero Positiva

El negocio presenta una rentabilidad neta limitada:

- **Revenue:** $47,195,297  

- **Profit Neto:** $928,281  

- **Margen Neto:** 2.0%

👉 **Conclusión:**

> ⚠️ El negocio es rentable, pero con un margen muy bajo, altamente sensible a cambios en la operación o en marketing.

---

## 🟥 2. Problema Crítico: Baja Efectividad Operativa

Distribución de pedidos:

- **Delivered:** 57.8%  
- **Cancelled:** 34.7%  
- **Returned:** 6.8%  

> ❗ El **42% de los pedidos no generan ingresos**, pero sí generan costos.


👉 A primera vista:

> ❗ Alto nivel de cancelaciones podría parecer ineficiencia

👉 Pero en realidad:

> 🧠 Las cancelaciones son parte de una **estrategia operativa deliberada** para evitar pérdidas mayores por devoluciones.

---

## 🔁 3. Rol Estratégico de las Cancelaciones

En el modelo COD:

- Un pedido **cancelado**:
  - ❌ No genera revenue
  - ❌ No incurre en costos logísticos
  - ✔ Solo pierde el costo de marketing

- Un pedido **devuelto**:
  - ❌ No genera revenue
  - ❌ Sí incurre en costos logísticos (envío + retorno)
  - ❌ Genera pérdida significativamente mayor

---

### 💣 Impacto financiero por tipo de pedido:

- Pedido entregado → **+ $29,576**

- Pedido cancelado → **- $21,195**

- Pedido devuelto → **- $94,107**

- Pedido con pérdida crítica → **- $98,748**

👉 **Insight clave:**

> 🔴 Una devolución puede costar hasta **4–5 veces más que una cancelación**

---

## 🧠 4. Insight Estratégico Principal

> 🔥 El negocio está optimizando pérdidas priorizando cancelaciones sobre devoluciones

👉 **Interpretación:**

- Las cancelaciones **no son completamente negativas**

- Funcionan como un **mecanismo de control de riesgo**

---

## ⚠️ 5. Problema Real: Calidad de la Demanda (Marketing + Oferta)

Aunque la estrategia de cancelación reduce pérdidas logísticas:

> ❗ Existe un problema más profundo: **la calidad de los pedidos generados**

---

### Evidencia:

- Alta proporción de pedidos cancelados (34.7%)
- Bajo margen neto (2%)
- Profit por pedido muy bajo ($1,439)

---

👉 **Insight clave:**

> ⚠️ El negocio está atrayendo una gran cantidad de clientes con baja intención real de compra

---

## 📣 6. Marketing: Genera volumen, pero no asegura calidad

- **ROAS:** 326%  

- **Marketing Cost:** $14.478.278  

👉 **Interpretación inicial:**
El marketing parece altamente rentable.

👉 **Pero en realidad:**

> ⚠️ El alto ROAS no se traduce en rentabilidad neta.

### ¿Por qué ocurre esto?

- El marketing genera pedidos, pero no controla su calidad
- Alta tasa de cancelaciones y fallos operativos
- Costos logísticos reducen significativamente el margen

👉 **Conclusión:**

> El marketing es eficiente en adquisición, pero ineficiente en generación de valor real.

---

## 🧠 7. Problema Estructural: Desalineación Oferta–Producto–Cliente

El problema no está solo en la operación, sino en la **propuesta de valor completa**:

---

### 🔍 Posibles causas:

#### 1. Oferta poco atractiva
- No genera urgencia o confianza suficiente

#### 2. Producto mal alineado con la necesidad
- No resuelve claramente un problema real

#### 3. Ángulo de venta débil o incorrecto
- No conecta con el público objetivo

#### 4. Expectativas mal comunicadas
- Diferencia entre lo que promete el anuncio y lo que recibe el cliente

---

👉 **Insight clave:**

> ❗ El marketing está trayendo volumen, pero no clientes de calidad

---

## 🟩 8. Unit Economics: Margen Insuficiente

Por pedido:

- **Average Order Value (AOV):** $73,171  

- **Costo operativo:** $49,285  

- **Marketing:** $22,447  

- **Profit:** $1,439  

👉 **Conclusión:**

> ⚠️ El margen es demasiado bajo para sostener errores operativos o mala calidad de demanda

---

## 🚚 9. Logística: Importante, pero no el problema principal

- **Shipping Cost:** $6,997,637  

- **Return Cost:** $976,411  

👉 Insight:

> La logística impacta la rentabilidad, pero el negocio ya está intentando controlarla mediante cancelaciones

---

👉 **Conclusión:**

> 🔸 La logística es un costo relevante  
> 🔸 Pero el problema raíz está en la generación de pedidos (demanda)

---

## 📉 10. Escalabilidad Riesgosa

👉 Aumentar volumen sin mejorar calidad implica:

- Más cancelaciones  
- Más devoluciones  
- Mayor presión sobre márgenes  

---

> ⚠️ Escalar marketing sin optimizar la conversión real puede destruir valor

---

## 🚀 11. Recomendaciones Estratégicas

### 🎯 1. Mejorar calidad de la demanda

- Refinar segmentación
- Ajustar campañas
- Optimizar targeting

---

### 🎯 2. Rediseñar la oferta

- Mayor claridad en beneficios
- Incentivos más fuertes
- Mejor propuesta de valor

---

### 🎯 3. Alinear producto–mercado

- Validar que el producto resuelve una necesidad real
- Ajustar portafolio si es necesario

---

### 🎯 4. Optimizar el ángulo de venta

- Mejor storytelling
- Comunicación clara
- Expectativas realistas

---

### 🎯 5. Mantener estrategia de cancelación (pero optimizarla)

- Seguir filtrando pedidos de alto riesgo
- Reducir cancelaciones mejorando calidad de leads

---

## 🧾 12. Conclusión Ejecutiva

> El negocio no presenta un problema puramente logístico, sino un desafío estructural en la calidad de la demanda generada. La alta proporción de cancelaciones es una decisión estratégica para evitar pérdidas mayores por devoluciones, lo que indica un intento de control de riesgo operativo. Sin embargo, el bajo margen neto y el alto volumen de pedidos no efectivos sugieren que el marketing está atrayendo clientes con baja intención real de compra. La principal oportunidad de mejora radica en optimizar la propuesta de valor, el ángulo de venta y la calidad del tráfico antes de escalar la inversión en marketing.

---

## ⭐ Insight Final

> 🔥 El crecimiento sostenible del negocio no depende de aumentar el tráfico, sino de mejorar la calidad de los pedidos generados.

> 👉 Mejor marketing NO significa más tráfico  
> 👉 Significa mejores clientes